### ISSUE TITLE
Deploy/retrieve individual components via metadata/tooling API

### ISSUE THREAD

**Issue by keirbowden** [forcedotcom/salesforcedx-vscode#488]
**Is your feature request related to a problem? Please describe.**
When not working against scratch orgs, moving code to/from target orgs requires use of the mdapi:deploy|retrieve sub commands. These are really heavyweight processes - most developers expect to be able to save/refresh individual files to/from the server quickly and the lack of this functionality hurts adoption imho

**Describe the solution you'd like **
Save a file saves to the Salesforce server, context menu for the component has a refresh from server option.

**Describe alternatives you've considered**
I’ve created a wrapper script around the mdapi commands that hides the details from the dev and integrated it as the default build tool. It’s a bit slow though.

**Additional context**
Force IDE/Mavens always had this feature -

**Comment by isalew** [comment]
A "prevent overwrites" option flag in the config file or extension settings would be a good idea too, in case you find yourself working out of shared environments.

**Comment by vazexqi** [comment]
@keirbowden The code is already in (see https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-vscode-core/package.json#L96) and it is guarded behind a preference. You would need the latest version (pre-release of the Salesforce CLI) and a new V44.0 org. Once Winter '19 has rolled out to everyone we will enable it by default.

**Comment by atnmorrison** [comment]
It would be great if there was an option in the convert command that only converted files that had been changed and automatically generated the package.xml so that only the changed components get deployed by the metadata api. Even better if we could combine the convert and deploy into one command I think we'd be set.

**Comment by atnmorrison** [comment]
Of course this all becomes pointless once / if all orgs support the sfdx version of metadata, is that in salesforce's plan or is the mdapi here to stay?

**Comment by vazexqi** [comment]
@atnmorrison - There are new CLI commands that do that. It's called `force:source:retrieve` and its counterpart is called `force:source:deploy`. It will do the auto-conversion for you.

More official information will be provided in the help docs soon. If you are feeling technical, you can try the pre-release release of the Salesforce CLI (requires an org that is already on Winter '19).  See https://www.npmjs.com/package/salesforcedx#install-a-tagged-version-of-salesforce-dx

**Comment by atnmorrison** [comment]
Right, I use those commands but it always converts everything in the force-app, it doesn't create a package.xml with only files that have been modified since the last conversion

**Comment by atnmorrison** [comment]
At any rate like it's still a PITA to move anything from new metadata api to old metadata. The dream is a single command to take a customer sandbox and create a personal scratch org that's shaped exactly like it, do all the dev in the scratch org then push only the committed files to the sandbox in a single command flagging any conflicts. Maybe it's all already there and I'm just doing it wrong which would be great! :)

**Comment by atnmorrison** [comment]
@vazexqi Awesome! The disclaimer makes it a none starter for anything other than side projects though. Looking forward to GA!
